docs/vm-metrics.md

```diff
@@ -0,0 +1,254 @@
+  even if the real data is not yet available.
+  Such metrics report the special value [TODO: which one? does this
+  concept exist in Kubernetes/Prometheus?] until the real data is available.
+  For example, the VM metric $VM_migration_data_total will be always
+  reported even before the first migration was attempted, if ever.
+
+- The node agent will have one option to keep the last known values of ephemeral metrics
+  after libvirt no longer reports them, for example again the $VM_migration_data_total metric
+  will be reported even after the migration is completed
+
+The normalization features will be controlled by node agent options (e.g. command line switches).
+
+### Metric naming rules
+Exposing the VM metrics to a prometheus server is pretty straightforward considering
+the [prometheus naming rules](https://prometheus.io/docs/practices/naming/).
+We will always one label to identify the VM who exposed the metric, for example
+```
+migration_memory_iteration{vm="$UUID"} 0
+vcpu_total_time{vm="$UUID",cpu="0"} 12345
+```
+
+### Implementation
+
+The requisites of the KubeVirt node agent, besides the obvious need to connect to libvirt and
+to expose a prometheus endpoint, are performance and reliability. The collection of metrics
+should require the minimal amount of resources, both as runtime and as container image.
+We know from the experience that shared storage issues may cause VMs to hang, and libvirt
+may in turn hang while collecting metrics from stuck VMs. We want the Node Agent to cope
+gracefully with this scenario, and keep reporting the metrics of the live VMs, skipping
+the stuck VMs. Reporting of stuck VMs is part of health monitoring, which is out of the
+scope of this document.
+
+The options to implement the KubeVirt node agent are:
+
+#### vAdvisor
+We can enhance the [vAdvisor project](https://github.com/KubeVirt/vAdvisor) to implement
+the missing features described above.
+
+**PROS**
+- The project already offers a fair amount of the features we need, it's a great starting point
+- Implemented in python, we can optionally reuse code from the oVirt monitoring
+
+**CONS**
+- Implemented in python, we know from experience in the oVirt project that python is a poor choice
+  scalability-wise. To properly address scale, we will need a rewrite in the future.
+- Implemented in python, requires a relatively big container image
+
+#### vAdvisor, reimplemented
+We can take vAdvisor as starting point and rewrite it in golang - to better integrate with
+the ecosystem right at the start instead of later, and implement the missing features on the
+golang codebase.
+
+**PROS**
+- No scalability issues
+- Better integration with the ecosystem
+- Minimal image size
+
+**CONS**
+- We will pay the rewrite cost upfront. But if we keep the design basics, we can minimize the effort.
+  [Initial experimentations](https://github.com/fromanirh/go-vAdvisor) gave promising results about
+  the development speed
+
+#### collectd
```

Comment:
I would try to focus on the mostly used tools which are afaik heapster, prometheus and cAdvisor. snapd and collectd seem to be mentioned quite regularly too. I don't have enough insight in real world deployments to see if snapd and collectd are important enough right now. Anyone?